File: app/src/main/java/com/dumuzeyn/mp3player/PersistedFolderStore.java
package com.dumuzeyn.mp3player;

import android.content.Context;
import android.content.Intent;
import android.content.UriPermission;
import android.net.Uri;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/** Compatibility boundary around SQLite-backed SAF music sources. */
final class PersistedFolderStore {
    private static final String PREFS = "voltune_music_folders";
    private static final String TREES = "trees";
    private static final String MIGRATED = "sources_migrated_to_sqlite";

    private PersistedFolderStore() {
    }

    static LibrarySource remember(Context context, Uri treeUri) {
        return remember(context, treeUri, fallbackName(treeUri), true);
    }

    static LibrarySource remember(Context context, Uri treeUri, String displayName,
            boolean explicitImport) {
        if (!hasReadPermission(context, treeUri)) {
            return null;
        }
        migrateLegacy(context);
        LibrarySourceStore store = new LibrarySourceStore(context);
        try {
            return store.remember(treeUri, displayName, explicitImport);
        } finally {
            store.close();
        }
    }

    static List<LibrarySource> list(Context context) {
        migrateLegacy(context);
        LibrarySourceStore store = new LibrarySourceStore(context);
        try {
            return store.list();
        } finally {
            store.close();
        }
    }

    static List<Uri> readableTrees(Context context) {
        ArrayList<Uri> result = new ArrayList<>();
        for (LibrarySource source : list(context)) {
            Uri uri = source.asUri();
            if (hasReadPermission(context, uri)) {
                result.add(uri);
            }
        }
        return result;
    }

    static boolean contains(Context context, Uri treeUri) {
        migrateLegacy(context);
        LibrarySourceStore store = new LibrarySourceStore(context);
        try {
            return store.find(treeUri) != null;
        } finally {
            store.close();
        }
    }

    static RemovedLibraryItems forget(Context context, LibrarySource source) {
        LibraryMutationStore store = new LibraryMutationStore(context);
        try {
            return store.removeSource(source.sourceId);
        } finally {
            store.close();
        }
    }

    static RemovedLibraryItems clear(Context context) {
        LibraryMutationStore store = new LibraryMutationStore(context);
        try {
            return store.clearLibrary();
        } finally {
            store.close();
        }
    }

    static void releaseReadPermission(Context context, Uri uri) {
        if (uri == null) {
            return;
        }
        try {
            context.getContentResolver().releasePersistableUriPermission(uri,
                    Intent.FLAG_GRANT_READ_URI_PERMISSION);
        } catch (SecurityException ignored) {
            // Permission may already have been revoked by Android or the provider.
        }
    }

    static boolean hasReadPermission(Context context, Uri uri) {
        if (uri == null) {
            return false;
        }
        for (UriPermission permission : context.getContentResolver()
                .getPersistedUriPermissions()) {
            if (permission.isReadPermission() && uri.equals(permission.getUri())) {
                return true;
            }
        }
        return false;
    }

    static int persistReadFlag(int resultFlags) {
        int flags = resultFlags & Intent.FLAG_GRANT_READ_URI_PERMISSION;
        return flags == 0 ? Intent.FLAG_GRANT_READ_URI_PERMISSION : flags;
    }

    private static void migrateLegacy(Context context) {
        android.content.SharedPreferences preferences = context.getSharedPreferences(PREFS,
                Context.MODE_PRIVATE);
        if (preferences.getBoolean(MIGRATED, false)) {
            return;
        }
        Set<String> legacy = new HashSet<>(preferences.getStringSet(TREES,
                new HashSet<String>()));
        LibrarySourceStore store = new LibrarySourceStore(context);
        try {
            for (String value : legacy) {
                Uri uri = Uri.parse(value);
                if (hasReadPermission(context, uri)) {
                    store.remember(uri, fallbackName(uri), false);
                }
            }
            preferences.edit().remove(TREES).putBoolean(MIGRATED, true).commit();
        } finally {
            store.close();
        }
    }

    private static String fallbackName(Uri uri) {
        String segment = uri == null ? "" : uri.getLastPathSegment();
        return segment == null || segment.trim().isEmpty() ? "Music folder" : segment;
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/PersistedFolderStore.kt
package com.dumuzeyn.mp3player

import android.content.Context
import android.content.Intent
import android.net.Uri

/** Compatibility boundary around SQLite-backed SAF music sources. */
object PersistedFolderStore {
    private const val PREFS = "voltune_music_folders"
    private const val TREES = "trees"
    private const val MIGRATED = "sources_migrated_to_sqlite"

    @JvmStatic
    fun remember(context: Context, treeUri: Uri): LibrarySource? =
        remember(context, treeUri, fallbackName(treeUri), true)

    @JvmStatic
    fun remember(
        context: Context,
        treeUri: Uri,
        displayName: String?,
        explicitImport: Boolean,
    ): LibrarySource? {
        if (!hasReadPermission(context, treeUri)) return null
        migrateLegacy(context)
        return LibrarySourceStore(context).use {
            it.remember(treeUri, displayName, explicitImport)
        }
    }

    @JvmStatic
    fun list(context: Context): List<LibrarySource> {
        migrateLegacy(context)
        return LibrarySourceStore(context).use { it.list() }
    }

    @JvmStatic
    fun readableTrees(context: Context): List<Uri> = list(context)
        .map(LibrarySource::asUri)
        .filter { hasReadPermission(context, it) }

    @JvmStatic
    fun contains(context: Context, treeUri: Uri): Boolean {
        migrateLegacy(context)
        return LibrarySourceStore(context).use { it.find(treeUri) != null }
    }

    @JvmStatic
    fun forget(context: Context, source: LibrarySource): RemovedLibraryItems =
        LibraryMutationStore(context).use { it.removeSource(source.sourceId) }

    @JvmStatic
    fun clear(context: Context): RemovedLibraryItems =
        LibraryMutationStore(context).use { it.clearLibrary() }

    @JvmStatic
    fun releaseReadPermission(context: Context, uri: Uri?) {
        if (uri == null) return
        try {
            context.contentResolver.releasePersistableUriPermission(
                uri,
                Intent.FLAG_GRANT_READ_URI_PERMISSION,
            )
        } catch (_: SecurityException) {
            // Permission may already have been revoked by Android or the provider.
        }
    }

    @JvmStatic
    fun hasReadPermission(context: Context, uri: Uri?): Boolean =
        uri != null && context.contentResolver.persistedUriPermissions.any {
            it.isReadPermission && it.uri == uri
        }

    @JvmStatic
    fun persistReadFlag(resultFlags: Int): Int {
        val flags = resultFlags and Intent.FLAG_GRANT_READ_URI_PERMISSION
        return flags.takeIf { it != 0 } ?: Intent.FLAG_GRANT_READ_URI_PERMISSION
    }

    private fun migrateLegacy(context: Context) {
        val preferences = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        if (preferences.getBoolean(MIGRATED, false)) return
        val legacy = HashSet(preferences.getStringSet(TREES, emptySet()).orEmpty())
        LibrarySourceStore(context).use { store ->
            legacy.forEach { value ->
                val uri = Uri.parse(value)
                if (hasReadPermission(context, uri)) {
                    store.remember(uri, fallbackName(uri), false)
                }
            }
            preferences.edit().remove(TREES).putBoolean(MIGRATED, true).commit()
        }
    }

    private fun fallbackName(uri: Uri?): String =
        uri?.lastPathSegment?.trim()?.takeIf(String::isNotEmpty) ?: "Music folder"
}
